src/shared/mount-util.c

```diff
@@ -1421,6 +1424,66 @@ int fd_make_mount_point(int fd) {
         return 1;
 }
 
+int mount_fd_clone(int mount_fd, bool recursive) {
+        int r;
+
+        assert(mount_fd >= 0);
+
+        /* Clone a detached mount (that may be owned by a foreign mountns, i.e. mountfsd's). For this we have
+         * to jump through some hoops, because the kernel currently doesn't allow us to just call
+         * open_tree(OPEN_TREE_CLONE) directly to get a clone of a mount that is detached and owned by
+         * another mountns. Hence here's what we do: we clone short-lived child in a new mount namespace
+         * owned by our userns. There, we attach the mount (invisible to anyone else). This is sufficient to
+         * pass the kernel check, so next we use open_tree(OPEN_TREE_CLONE) to get our own detached mount.
+         * This we send back to the parent, which then can use it. */
+
+        _cleanup_close_pair_ int transfer_fds[2] = EBADF_PAIR;
+        r = socketpair(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0, transfer_fds);
+        if (r < 0)
+                return log_debug_errno(errno, "Failed to create socket pair: %m");
+
+        _cleanup_close_pair_ int errno_pipe_fds[2] = EBADF_PAIR;
+        if (pipe2(errno_pipe_fds, O_CLOEXEC|O_NONBLOCK) < 0)
+                return log_debug_errno(errno, "Failed to open pipe: %m");
+
+        /* Fork a child. Note that we set FORK_NEW_MOUNTNS|FORK_MOUNTNS_SLAVE here, i.e. get a new mount namespace */
+        r = safe_fork_full(
+                        "(clonemnt)",
+                        /* stdio_fds= */ NULL,
+                        (int[]) { mount_fd, transfer_fds[1], errno_pipe_fds[1] }, 3,
+                        FORK_RESET_SIGNALS|FORK_CLOSE_ALL_FDS|FORK_DEATHSIG_SIGTERM|FORK_LOG|FORK_REOPEN_LOG|FORK_WAIT|FORK_NEW_MOUNTNS|FORK_MOUNTNS_SLAVE,
```

Comment:
This is an internal helper and does not alter visible state of the system, so might as well use SIGKILL?